### ISSUE TITLE
📊 Enhance Sum Operation with Axis and Keepdim Support

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#17]
# 📊 Enhance Sum Operation with Axis and Keepdim Support

## 📋 Description

The current `Tensor.sum()` implementation is basic and only supports summing all elements. We need to enhance it to support axis-specific reduction and keepdim functionality, matching PyTorch/NumPy behavior.

## 🎯 Acceptance Criteria

- [ ] Add `axis` parameter to specify reduction dimension(s)
- [ ] Add `keepdim` parameter to preserve reduced dimensions
- [ ] Support multiple axes (e.g., `axis=(0, 2)`)
- [ ] Support negative axis indexing
- [ ] Implement proper gradient computation for axis-specific sums
- [ ] Handle empty tensors and edge cases
- [ ] Add comprehensive tests for all axis combinations
- [ ] Ensure memory efficiency for large tensor reductions
- [ ] Support all data types with proper casting

### Current State
- `Tensor.sum()`: ✅ Basic implementation (sums all elements)
- Axis support: ❌ Not implemented
- Keepdim support: ❌ Not implemented
- Gradient computation: ✅ Basic (requires enhancement)

### Enhanced Implementation

**Method Signature**
```python
@staticmethod
def sum(
    t: Tensor,
    axis: Optional[Union[int, Tuple[int, ...]]] = None,
    keepdim: bool = False,
    dtype: Optional[DType] = None,
) -> Tensor:
    """
    Sum tensor elements along specified axis/axes.
    
    Args:
        t: Input tensor
        axis: Axis or axes to reduce. None means all axes.
        keepdim: Whether to keep reduced dimensions as size 1
        dtype: Output data type
    """
```

**Implementation Strategy**
```python
# Handle axis parameter
if axis is None:
    # Sum all elements (current behavior)
    pass
elif isinstance(axis, int):
    # Single axis reduction
    axis = axis % len(t.shape)  # Handle negative indexing
elif isinstance(axis, tuple):
    # Multiple axes reduction
    axis = tuple(a % len(t.shape) for a in axis)

# Calculate output shape
output_shape = list(t.shape)
if keepdim:
    for ax in sorted(axis, reverse=True):
        output_shape[ax] = 1
else:
    for ax in sorted(axis, reverse=True):
        del output_shape[ax]
```

### Autograd Support

**SumAxis Function Class**
```python
class SumAxis(Function):
    @staticmethod
    def forward(ctx, input_tensor, axis, keepdim):
        ctx.input_shape = input_tensor.shape
        ctx.axis = axis
        ctx.keepdim = keepdim
        return sum_along_axis(input_tensor, axis, keepdim)
    
    @staticmethod
    def backward(ctx, grad_output):
        # Expand gradient back to original shape
        if not ctx.keepdim and ctx.axis is not None:
            # Add back reduced dimensions
            grad_output = expand_dims(grad_output, ctx.axis)
        
        # Broadcast to original shape
        return broadcast_to(grad_output, ctx.input_shape)
```

## 🧪 Testing Requirements

- [ ] Test single axis reduction: `tensor.sum(axis=0)`
- [ ] Test multiple axes: `tensor.sum(axis=(0, 2))`
- [ ] Test keepdim behavior: `tensor.sum(axis=1, keepdim=True)`
- [ ] Test negative axis indexing: `tensor.sum(axis=-1)`
- [ ] Test with different tensor shapes (1D, 2D, 3D+)
- [ ] Test gradient computation for axis-specific sums
- [ ] Test edge cases: empty tensors, single-element tensors
- [ ] Test dtype casting behavior
- [ ] Performance benchmarks vs NumPy
- [ ] Memory usage profiling for large tensors

## 📁 Files to Modify

- `grad/tensor.py` - Enhance sum method
- `grad/autograd/ops.py` - Add SumAxis function class
- `grad/utils/reduction.py` - New file for reduction utilities
- `tests/test_sum_enhanced.py` - Comprehensive test suite
- `tests/test_reduction_gradients.py` - Gradient verification tests

## 🔗 Dependencies

- Requires: Broadcasting system (#02)
- Requires: Autograd backward pass (#01)
- Uses: Existing tensor indexing and stride system
- Blocks: Mean operation (#06)
- Blocks: Other reduction operations (#06)

### Basic Usage
```python
# 3D tensor example
x = Tensor.randn(2, 3, 4, requires_grad=True)

# Sum all elements (current behavior)
total = x.sum()  # Shape: ()

# Sum along specific axis
sum_axis0 = x.sum(axis=0)     # Shape: (3, 4)
sum_axis1 = x.sum(axis=1)     # Shape: (2, 4)
sum_axis2 = x.sum(axis=2)     # Shape: (2, 3)

# Sum along multiple axes
sum_multi = x.sum(axis=(0, 2))  # Shape: (3,)

# Keep dimensions
sum_keepdim = x.sum(axis=1, keepdim=True)  # Shape: (2, 1, 4)

# Negative indexing
sum_last = x.sum(axis=-1)     # Same as axis=2
```

### Gradient Computation
```python
x = Tensor([[1, 2, 3], [4, 5, 6]], requires_grad=True)
y = x.sum(axis=0)  # [5, 7, 9]
loss = y.sum()     # 21

loss.backward()
print(x.grad)      # [[1, 1, 1], [1, 1, 1]]
```

### Advanced Usage
```python
# Neural network layer example
batch_size, features = 32, 128
hidden = Tensor.randn(batch_size, features, requires_grad=True)

# Compute mean by summing and dividing
feature_sum = hidden.sum(axis=0, keepdim=True)  # (1, 128)
feature_mean = feature_sum / batch_size

# Loss computation
loss = (feature_mean ** 2).sum()
loss.backward()
```

## 🚨 Edge Cases to Handle

- Empty tensors with different reduction axes
- Single-element tensors
- Reduction along all axes vs axis=None
- Invalid axis values (out of bounds)
- Duplicate axes in tuple
- Very large tensors (memory constraints)
- Mixed data types with dtype parameter
- Gradient shape broadcasting edge cases

## 🔧 Performance Considerations

- Efficient memory access patterns for different axes
- Minimize temporary memory allocation
- Optimize for contiguous vs non-contiguous tensors
- Consider cache-friendly reduction algorithms
- Memory pooling for intermediate results
- Vectorized operations where possible

## 🏷️ Labels

- `enhancement`
- `reduction-operations`
- `core-functionality`
- `medium-priority`

## 📈 Priority

**Medium-High** - Essential for implementing mean() and other reductions

## 🔗 Related Issues

- Depends on: Autograd system (#01)
- Depends on: Broadcasting (#02)
- Blocks: Mean operation (#06)
- Blocks: Statistical operations
- Related to: Tensor shape manipulation
